### ISSUE TITLE
Vec::as_non_empty_slice let's user choose arbitrary lifetime

### ISSUE THREAD

**Issue by 197g** [lzutao/rust-oom#2]
The impl block of NonEmptyVec introduces a _new_ lifetime that is chosen for the output type of `NonEmptySlice` etc. However, that means it's an impl *for all* lifetimes and the caller is free to chose one. It should always return the implicit `&self` lifetime, i.e. use `'_` instead of a new one.

```rust
let vec = oom::NonEmptyVec::<u8>::from_vec(vec![0]);
let bad: oom::NonEmptySlice<'static, _> = vec.as_nonempty_slice();
drop(vec);
let dangling = bad.first();
```

**Comment by 197g** [comment]
The same is true for the impl blocks of NonEmptySlice and NonEmptySliceMut however there it is only critical for methods that return `&'a mut T`. That let's one create aliasing pointers which is also UB.

**Comment by tesuji** [comment]
Thank you for reporting this. I want to add some tests for this. Do you how could I make compile-fail tests?

**Comment by 197g** [comment]
`compile_fail` tests can only be done as doctests: https://doc.rust-lang.org/rustdoc/documentation-tests.html#attributes

**Comment by tesuji** [comment]
Thank. But adding compile-fail doctest to public API is bad ? Do you have any suggestions ?

**Comment by 197g** [comment]
You can add doctests on empty extern blocks:

```rust
/// ```compile_fail
/// shouldn't compile
/// ```
extern {}
```

**Comment by tesuji** [comment]
Nice.

**Comment by tesuji** [comment]
Thank you again. This should be fixed and tested in d29043c4602c4fa347b30cd3c7ef355dabb08188 .

**Comment by tesuji** [comment]
> the impl blocks of NonEmptySlice and NonEmptySliceMut

Actually, I don't understand what's that about ? Can you take a look again ? Thanks.

**Comment by 197g** [comment]
The methods of `NonEmptyMutSlice` allow you to create references with the original lifetime. That is unsound since it violates the uniqueness guarantee of `&mut` references. This should never compile:

```rust
let mut t = 0;
let mut slice = oom::NonEmptyMutSlice::<u8>::from_mut(&mut t);
let first = slice.first_mut();
let second = slice.first_mut();
assert_eq!(first as *mut _, second as *mut _);
*first = 0;
*second = 1;
assert_eq!(*first, 1, "UB, written here through second.");
```

**Comment by tesuji** [comment]
Thank you very much, I had a misconception about how `impl<'a>` work, I thought
`&self -> &'a` is equivalent to `&'a self -> &'a`. Clearly I was wrong. This should be fixed by 2c0e3da.
I rewrote the entire code so that it just know a wrapper around `&[T]` and `Vec<T>`.
I am really appreciate the time you took to review the code.
